File: app/src/main/java/com/esorokin/lantector/ui/plugins/NavigationPlugin.java
package com.esorokin.lantector.ui.plugins;

import android.support.v7.app.AppCompatActivity;

import com.art.alligator.NavigationContext;
import com.art.alligator.NavigationContextBinder;
import com.art.alligator.Navigator;
import com.art.alligator.TransitionAnimationProvider;
import com.esorokin.lantector.di.DependencyManager;
import com.esorokin.lantector.presentation.navigation.animation.DefaultAnimationProvider;
import com.esorokin.lantector.ui.plugins.base.BasePlugin;

import javax.inject.Inject;

public class NavigationPlugin extends BasePlugin {
	@Inject
	NavigationContextBinder navigationContextBinder;

	@Inject
	Navigator navigator;

	private AppCompatActivity activity;
	private TransitionAnimationProvider transitionAnimationProvider;

	public NavigationPlugin(AppCompatActivity activity) {
		this(activity, new DefaultAnimationProvider());
	}

	public NavigationPlugin(AppCompatActivity activity, TransitionAnimationProvider transitionAnimationProvider) {
		super();
		DependencyManager.getAppComponent().inject(this);
		this.activity = activity;
		this.transitionAnimationProvider = transitionAnimationProvider;
	}

	@Override
	public void onResume() {
		super.onResume();
		navigationContextBinder.bind(new NavigationContext.Builder(activity)
				.transitionAnimationProvider(transitionAnimationProvider)
				.build());
	}

	@Override
	public void onPause() {
		super.onPause();
		navigationContextBinder.unbind();
	}

	@Override
	public void onBackPressed() {
		super.onBackPressed();
		navigator.goBack();
	}
}

File: app/src/main/java/com/esorokin/lantector/ui/plugin/NavigationPlugin.kt
package com.esorokin.lantector.ui.plugin

import android.support.v7.app.AppCompatActivity

import com.art.alligator.NavigationContext
import com.art.alligator.NavigationContextBinder
import com.art.alligator.Navigator
import com.art.alligator.TransitionAnimationProvider
import com.esorokin.lantector.di.DependencyManager
import com.esorokin.lantector.presentation.navigation.animation.DefaultAnimationProvider
import com.esorokin.lantector.ui.plugin.base.BasePlugin

import javax.inject.Inject

class NavigationPlugin(private val activity: AppCompatActivity,
                       private val transitionAnimationProvider: TransitionAnimationProvider = DefaultAnimationProvider()) : BasePlugin() {
    @Inject
    internal lateinit var navigationContextBinder: NavigationContextBinder

    @Inject
    internal lateinit var navigator: Navigator

    init {
        DependencyManager.appComponent.inject(this)
    }

    override fun onResume() {
        super.onResume()
        navigationContextBinder.bind(NavigationContext.Builder(activity)
                .transitionAnimationProvider(transitionAnimationProvider)
                .build())
    }

    override fun onPause() {
        super.onPause()
        navigationContextBinder.unbind()
    }

    override fun onBackPressed() {
        super.onBackPressed()
        navigator.goBack()
    }
}
